kore/src/Kore/Strategies/Goal.hs

```diff
@@ -921,39 +938,88 @@ deriveSeq rules goal = errorBracket $ do
  -}
 removalPredicate
     :: SimplifierVariable variable
+    => MonadSimplify m
     => Pattern variable
     -- ^ Destination
     -> Pattern variable
     -- ^ Current configuration
-    -> Predicate variable
-removalPredicate destination config =
-    let
-        -- The variables of the destination that are missing from the
-        -- configuration. These are the variables which should be existentially
-        -- quantified in the removal predicate.
-        configVariables =
-            Attribute.FreeVariables.getFreeVariables
-            $ Pattern.freeVariables config
-        destinationVariables =
-            Attribute.FreeVariables.getFreeVariables
-            $ Pattern.freeVariables destination
-        extraVariables = Set.toList
-            $ Set.difference destinationVariables configVariables
-        extraElementVariables = [v | ElemVar v <- extraVariables]
-        extraNonElemVariables = filter (not . isElemVar) extraVariables
-        quantifyPredicate =
-            Predicate.makeMultipleExists extraElementVariables
-    in
-        if not (null extraNonElemVariables)
-        then error
-            ("Cannot quantify non-element variables: "
-                ++ show (unparse <$> extraNonElemVariables))
-        else Predicate.makeNotPredicate
-            $ quantifyPredicate
-            $ Predicate.makeCeilPredicate_
-            $ mkAnd
-                (Pattern.toTermLike destination)
-                (Pattern.toTermLike config)
+    -> m (Predicate variable)
+removalPredicate
+    destination
+    configuration
+  | isFunctionPattern configTerm
+  , isFunctionPattern destTerm
+  = do
+    unifiedConfigs <- simplifyTerm (mkAnd configTerm destTerm)
+    if OrPattern.isFalse unifiedConfigs
+        then return Predicate.makeTruePredicate_
+        else
+            case Foldable.toList unifiedConfigs of
+                [substPattern] ->
+                    let extraNonElemVariables =
+                            remainderNonElemVariables configuration substPattern
+                        extraElemVariables =
+                            remainderElementVariables configuration substPattern
+                    in if not (null extraNonElemVariables)
+                        then
+                            error . show . Pretty.vsep $
+                                "Cannot quantify non-element variables: "
+                                : fmap (Pretty.indent 4 . unparse) extraNonElemVariables
```

Comment:
Let's factor this branch out into a function which returns `extraElemVariables` or throws an error if there are non-element variables. Then, we only need to write the "happy path" here.